### PR TITLE
Rewrite to use ScreenOrientation interface (closes #13)

### LINKED ISSUE #13: OrientationInformation should be a dictionary
It is not a constructible class, and has no methods or `this`-dependent constructs, and so should just be a normal JavaScript object (i.e. "dictionary" in WebIDL terms).

### PR DESCRIPTION
**Pull request by marcoscaceres**
- added ScreenOrientation interface
- Rewrote examples
- Removed redundant "dependencies"
- Merged dependencies and terminology section
- Made Promise linkable in dependencies
- Moved ScreenOrientation and Extension to navigator object to the top
  of the document
- Removed "concepts" section (left all inner content, of course)
- rewrote how to get attributes
- removed redundancy when invoking lock() and unlock()
- Moved enums to their  own sections (it was annoying trying to find
  them. Now the show up in the TOC!)
- Introduced "security condition" concept. This fixes the weird
  dependency on the FULLSCREEN API. That's now moved to the "Interaction
  with the FullScreen API" section
-  Added Interaction with FullScreen API section. Defined what to do
  when full screen fully exits there.
- moved conformance section to bottom of document

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -66,49 +66,31 @@
...
-        &lt;!DOCTYPE html&gt;
-        &lt;html&gt;
-          &lt;head&gt;
-            &lt;title&gt;Screen Orientation API Example 1&lt;/title&gt;
-            &lt;script&gt;
-              function show() {
-                alert("Screen orientation state is now " + screen.orientation);
-              }
+&lt;script&gt;
+  var show = () =&gt; {
```
Comment (mounirlamouri) [inline index.html]: Maybe we should avoid fancy syntax and simply have:

``` js
function show() {
  ...
}
```

File: index.html [hunk 2]
```diff
@@ -66,49 +66,31 @@
...
-            &lt;input type='button' value='Unlock'
-              onclick='screen.unlockOrientation();'&gt;
-            &lt;input type='button' value='Lock to portrait'
-              onclick="screen.lockOrientation('portrait');"&gt;
-            &lt;input type='button' value='Lock to landscape'
-              onclick="screen.lockOrientation('landscape');"&gt;
-          &lt;/body&gt;
-        &lt;/html&gt;
+&lt;button onclick='screen.orientation.unlock()'&gt;
+  Unlock
```
Comment (mounirlamouri) [inline index.html:84]: You forgot `</button>`.

File: index.html [hunk 3]
```diff
@@ -66,49 +66,31 @@
...
 
+&lt;button onclick="screen.orientaiton.lock('portrait')"&gt;
+  Lock to portrait
```
Comment (mounirlamouri) [inline index.html:87]: ditto

File: index.html [hunk 4]
```diff
@@ -66,49 +66,31 @@
...
+
+&lt;button onclick="screen.orientation.lock('landscape')"&gt;
+  Lock to landscape
```
Comment (mounirlamouri) [inline index.html:90]: ditto

File: index.html [hunk 5]
```diff
@@ -123,25 +105,10 @@
...
-        </li>
-        <li>
-          <dfn><a href=
-          "http://dev.w3.org/html5/spec/webappapis.html#event-handler-event-type">
-          event handler event type</a></dfn>
-        </li>
-        <li>
-          <dfn><a href=
-          '%20http://www.w3.org/TR/html5/webappapis.html#event-handler-idl-attributes'>
+          "www.w3.org/html/wg/drafts/html/master/webappapis.html#event-handler-idl-attributes">
```
Comment (mounirlamouri) [inline index.html]: You forgot `http://...`

File: index.html (hunk 1 above)
Comment (marcoscaceres) [inline index.html]: reverted.

File: index.html (hunk 4 above)
Comment (marcoscaceres) [inline index.html:90]: button is self closing, but I added the closing tag back in

File: index.html (hunk 5 above)
Comment (marcoscaceres) [inline index.html]: fixed

File: index.html [hunk 6]
```diff
@@ -199,418 +153,18 @@
...
         fullscreen</a></dfn> are defined in [[!FULLSCREEN]].
       </p>
-    </section>
-    <section>
-      <h2>
-        Concepts
-      </h2>
-      <section>
-        <h2>
-          Terminology
```
Comment (mounirlamouri) [inline index.html:209]: I added this section for `screen` on purpose. I didn't want this to be merged with Dependencies because it is not a dependency, it's a terminology we use in the document.

File: index.html [hunk 7]
```diff
@@ -619,58 +173,435 @@
       [[!CSSOM-VIEW]], which this specification extends:
       <dl class='idl' title='partial interface Screen'>
         <dt>
-          readonly attribute OrientationInformation orientation
+          readonly attribute ScreenOrientation orientation
+        </dt>
+      </dl>
+    </section>
+    <section>
```
Comment (mounirlamouri) [inline index.html]: What about having all the WebIDL sections into a `Interface definitions` section?

File: index.html [hunk 8]
```diff
@@ -619,58 +173,435 @@
...
+        return the <a>current screen orientation angle</a>.
+      </p>
+      <p>
+        The <dfn>onchange</dfn> attribute is an event handler IDL attribute.
       </p>
     </section>
     <section>
       <h2>
-        Extensions to the <a>WindowEventHandlers</a> interface
+        enum <code>OrientationType</code>
```
Comment (mounirlamouri) [inline index.html]: &lt;code>OrientationType&lt;/code> enum

File: index.html [hunk 9]
```diff
@@ -619,58 +173,435 @@
...
+          landscape-primary
+        </dt>
+        <dt>
+          landscape-secondary
         </dt>
       </dl>
+    </section>
+    <section>
+      <h2>
+        enum <code>OrientationLockType</code>
```
Comment (mounirlamouri) [inline index.html]: &lt;code>OrientationLockType&lt;/code> enum

Comment (marcoscaceres) [inline index.html]: fixed

File: index.html (hunk 8 above)
Comment (marcoscaceres) [inline index.html]: fixed

File: index.html [hunk 10]
```diff
@@ -619,58 +173,435 @@
...
+          landscape-primary
+        </dt>
+        <dt>
+          landscape-secondary
+        </dt>
+      </dl>
+    </section>
+    <section>
+      <h2>
+        Reading the screen orientation
```
Comment (mounirlamouri) [inline index.html]: Would you be against having &lt;section>Concepts&lt;/code> to wrap all the concepts? I think it makes the TOC more readable.

Comment (marcoscaceres) [inline index.html]: I'm ok with it. Will add it.

File: index.html (hunk 7 above)
Comment (marcoscaceres) [inline index.html]: I don't think you need it (it's not great having sections that are just empty wrappers). But what the hell, we did it for Concepts already so I'll add it :dancer:. Done!

File: index.html [hunk 11]
```diff
@@ -227,14 +204,93 @@
...
+          When getting the <code><dfn id=
+          "widl-ScreenOrientation-type">type</dfn></code> attribute, the user
+          agent MUST return the <a>current screen orientation type</a>.
         </p>
         <p>
-          The <code>angle</code> attribute MUST return the <a>current screen
-          orientation angle</a>.
+          When getting the <code><dfn id=
+          "widl-ScreenOrientation-angle">angle</dfn></code> attribute, the use
+          agent MUST return the <a>current screen orientation angle</a>.
```
Comment (mounirlamouri) [inline index.html]: typo: user agent

File: index.html [hunk 12]
```diff
@@ -310,8 +352,10 @@
         </h2>
         <p>
           The <a>user agent</a> MAY require a <a>document</a>'s <a>top-level
-          browsing context</a> to be fullscreen [[!FULLSCREEN]] in order to be
-          able to lock the screen orientation.
+          browsing context</a> meet one or more <dfn>security conditions</dfn>
+          order to be able to lock the screen orientation. For example, a use
+          agent might require the top-level browsing context to be fullscreen
```
Comment (mounirlamouri) [inline index.html]: typo: user agent

File: index.html [hunk 13]
```diff
@@ -310,8 +352,10 @@
...
+          [[FULLSCREEN]] in order to allow an orientation change.
```
Comment (mounirlamouri) [inline index.html]: in order to allow an orientation **lock**.

File: index.html [hunk 14]
```diff
@@ -716,50 +670,87 @@
...
+        <p>
+          As a <a title="security conditions">security condition</a>, a user
+          agent MAY restrict locking the screen orientation exclusively to when
+          the top-level browsing document's full-screen enabled flag has been
+          set [[!FULLSCREEN]] (i.e., when the document is in fullscreen).
+        </p>
+        <p>
+          If the screen is locked to some orientation when in fullscreen, then,
+          when the document fully exits fullscreen, the user agent MUST lock
+          the orientation to the default orientation.
```
Comment (mounirlamouri) [inline index.html]: Sure. But only if the lock is possible because the document was fullscreen.

File: index.html [hunk 15]
```diff
@@ -716,50 +670,87 @@
...
+          game pointing to an objective.
+        </p>
+        <p>
+          Drawing on the screen in order to point to a physical location
+          requires to know the device orientation and the orientation of the
+          screen in the device coordinates. Without the APIs defined in this
+          specification, a developer has to assume that the <a>current screen
+          orientation angle</a> is 0. With the help of the APIs described in
+          this specification, the developer can <a>apply an orientation
+          lock</a> using <code>0</code> to make that assumption a certitude.
```
Comment (mounirlamouri) [inline index.html]: I know you didn't really changed that, but could you:
s/using &lt;code>0&lt;/code>/using &lt;code>natural&lt;/code>/

File: index.html [hunk 16]
```diff
@@ -227,14 +204,93 @@
...
         </p>
         <p>
+          The <code><dfn id=
+          "widl-ScreenOrientation-onchange">onchange</dfn></code> attribute is
+          an <a>event handler IDL attribute</a>.
+        </p>
+      </section>
+      <section>
+        <h2>
+          enum <code>OrientationType</code>
```
Comment (mounirlamouri) [inline index.html]: Hmm, did you revert the change you did before?
&lt;code>Foo&lt;/code> enum
instead of
enum &lt;code>Foo&lt;/code>

File: index.html [hunk 17]
```diff
@@ -227,14 +204,93 @@
...
+            landscape-primary
+          </dt>
+          <dt>
+            landscape-secondary
+          </dt>
+        </dl>
+      </section>
+      <section>
+        <h2>
+          enum <code>OrientationLockType</code>
```
Comment (mounirlamouri) [inline index.html]: ditto

File: index.html [hunk 18]
```diff
@@ -66,49 +66,32 @@
...
-      </p>
-      <p>
-        Implementations that use ECMAScript to expose the APIs defined in this
-        specification MUST implement them in a manner consistent with the
-        ECMAScript Bindings defined in the Web IDL specification [[!WEBIDL]].
-      </p>
-    </section>
     <section>
       <h2>
         Dependencies
```
Comment (mounirlamouri) [inline index.html:108]: I'm still not a big fan of that merge. Can we keep the dependencies and the "terms" in different places?

File: index.html [hunk 19]
```diff
@@ -199,26 +154,48 @@
         'http://www.w3.org/TR/fullscreen/#fully-exit-fullscreen'>fully exit
         fullscreen</a></dfn> are defined in [[!FULLSCREEN]].
```
Comment (mounirlamouri) [inline index.html:177]: As long as you remove dependencies, we might want to remove FULLSCREEN from the dependencies. It is only mentioned in non normative requirements so lets not mention it outside so it will be a non-normative dependency. WDYT?

File: index.html [hunk 20]
```diff
@@ -199,26 +154,48 @@
...
       </p>
+      <p>
+        Unless stated otherwise, the term <dfn>screen</dfn> is equivalent to
+        the screen of the output device associated to the <a>Window</a>, as per
+        [[!CSSOM-VIEW]].
+      </p>
+      <p>
+        Algorithms defined in this specification assume that for each
+        <a>top-level browsing context</a> there is a <dfn>pending
+        promise</dfn>, which is initially set to <code>null</code>, which is a
```
Comment (mounirlamouri) [inline index.html]: You are doing <dfn>pending-promise</dfn> but you never actually link to it. Is that on purpose?

File: index.html [hunk 21]
```diff
@@ -227,14 +204,93 @@
...
-          The <code>angle</code> attribute MUST return the <a>current screen
-          orientation angle</a>.
+          When getting the <code><dfn id=
+          "widl-ScreenOrientation-angle">angle</dfn></code> attribute, the use
+          agent MUST return the <a>current screen orientation angle</a>.
         </p>
         <p>
+          The <code><dfn id=
+          "widl-ScreenOrientation-onchange">onchange</dfn></code> attribute is
+          an <a>event handler IDL attribute</a>.
```
Comment (mounirlamouri) [inline index.html]: Hmm, I don't think that's enough. It seems that other specs would do something like:

```
<code><dfn id='widl-ScreenOrientation-onchange'>onchange</dfn></code> is an <a>event handler</a> with the corresponding <code>change</code>
<a>event handler event type</a> that MUST be supported as attribute solely by the
<a>ScreenOrientation</a> object.
```

If you do that, you will have to update the dependencies.

File: index.html [hunk 22]
```diff
@@ -575,22 +582,30 @@
...
+          visible per [[!PAGE-VISIBILITY]], abort these steps.
           </li>
-          <li>If a Promise created in <a>apply an orientation lock</a> steps is
-          pending, it MUST be resolved with <code>undefined</code>.
-          </li>
-          <li>The <a>user agent</a> MUST <a>queue a task</a> to <a>fire a
-          simple event</a> named <code>orientationchange</code> at the
-          <a>browsing context</a> <a>active document</a>'s <a>Window</a> object
-          and wait until that task begins to continue to the next step.
+          <li>If <a>pending promise</a> is not <code>null</code>:
```
Comment (mounirlamouri) [inline index.html]: We should update the pending promise _after_ we update the current angle/type.

File: index.html [hunk 23]
```diff
@@ -575,22 +582,30 @@
...
+            <ol>
+              <li>Resolve <a>pending promise</a> with <code>undefined</code>.
```
Comment (mounirlamouri) [inline index.html]: Not _always_. It should only happen if the new orientation matches the lock.

File: index.html [hunk 24]
```diff
@@ -575,22 +582,30 @@
...
+              </li>
+              <li>Update the <a>current screen orientation angle</a>.
+              </li>
+              <li>
+                <a>Fire a simple event</a> named <code>change</code> at each
+                <a href=
+                "#widl-Screen-orientation"><code>screen.orientation</code></a>
+                object.
+              </li>
+            </ol>
```
Comment (mounirlamouri) [inline index.html:627]: Could you add a paragraph making it clear that this mean that a non-visible page will get stalled orientation values which will be updated when the page will be visible again - with a change event.

File: index.html [hunk 25]
```diff
@@ -310,8 +352,10 @@
         </h2>
         <p>
           The <a>user agent</a> MAY require a <a>document</a>'s <a>top-level
-          browsing context</a> to be fullscreen [[!FULLSCREEN]] in order to be
-          able to lock the screen orientation.
+          browsing context</a> meet one or more <dfn>security conditions</dfn>
+          order to be able to lock the screen orientation. For example, a use
```
Comment (mounirlamouri) [inline index.html]: typo: in order

File: index.html (hunk 18 above)
Comment (marcoscaceres) [inline index.html:108]: Sure, will split it up again. But do take a look after. It means you need to go to two different places to find essentially the same thing (definitions and links to definitions).

File: index.html (hunk 20 above)
Comment (marcoscaceres) [inline index.html]: Hmm... I thought I did link to it. That's a bug.

File: index.html (hunk 11 above)
Comment (marcoscaceres) [inline index.html]: fixed

File: index.html (hunk 21 above)
Comment (marcoscaceres) [inline index.html]: I can add that, but which spec are you basing that on? 

you really don't need the "that MUST be supported as attribute solely by the <a>ScreenOrientation</a> object." The WebIDL already says that.

File: index.html (hunk 16 above)
Comment (marcoscaceres) [inline index.html]: heh, I might have "fixed" the one I fixed! :tada:

File: index.html (hunk 12 above)
Comment (marcoscaceres) [inline index.html]: fixed

File: index.html (hunk 13 above)
Comment (marcoscaceres) [inline index.html]: fixed

File: index.html (hunk 15 above)
Comment (marcoscaceres) [inline index.html]: Added natural

**Comment by mounirlamouri** [comment]
Crap, GitHub doesn't save draft comments... :(

I have a few general comments.
- It seems that you usually don't linkify `user agent` but the document usually does. With your changes, it becomes fairly inconsistent. Could you keep the `<a>` around it? Or have a commit to remove all of them. I don't care much but I would like consistency :)
- It bothers me a bit that we return `SecurityError` for fullscreen and sandbox. It might make it hard to understand what's happening but also, the fullscreen requirement is more a UX issue than a security error. I don't see a better solution though.
- Why moved conformance at the bottom? It is usually at the beginning and I'm not sure what's the benefit of moving it at the bottom.

I also have a couple of comments from your initial list of changes:
- rewrote how to get attributes
  -> what did you do exactly?
- removed redundancy when invoking lock() and unlock()
  -> what did you do exactly?

Finally, could you add your name after mine in the author list?

File: index.html (hunk 21 above)
Comment (mounirlamouri) [inline index.html]: http://xhr.spec.whatwg.org

File: index.html (hunk 18 above)
Comment (mounirlamouri) [inline index.html:108]: I have an idea. What about adding the non-dependency things at the beginning of "Concepts".
Like:

```
<section>
<h2>Concepts</h2>
When the spec says <b>screen</b> it means unicorn.
For every top level browsing window, this specification has a pending kitten.
<section>
<h2>Reading the screen orientation</h2>
[...]
```

Comment (marcoscaceres) [inline index.html:108]: like it! :+1:

File: index.html (hunk 21 above)
Comment (marcoscaceres) [inline index.html]: You've been reading **Van** Kesteren! I should have known!

Comment (marcoscaceres) [inline index.html]: Updated it to this:

``` HTML
          The <code><dfn id=
          "widl-ScreenOrientation-onchange">onchange</dfn></code> attribute is
          an <a>event handler</a> whose corresponding <a>event handler event type</a> 
          is <code>change</code>.
```

That matches what Anne's spec has, includes the name of the event handler event type, without having to repeat the information that is already in the WebIDL.

File: index.html (hunk 25 above)
Comment (marcoscaceres) [inline index.html]: fixed

File: index.html (hunk 21 above)
Comment (mounirlamouri) [inline index.html]: Sounds good. Make sure to add the &lt;dfn> in the Definitions section.

File: index.html (hunk 20 above)
Comment (marcoscaceres) [inline index.html]: linked up all "user agents" now.

File: index.html (hunk 21 above)
Comment (marcoscaceres) [inline index.html]: Yep, added definitions. Thanks for the reminder.

File: index.html (hunk 24 above)
Comment (marcoscaceres) [inline index.html:627]: We should do this as a separate bug.

Comment (marcoscaceres) [inline index.html:627]: Created #28

File: index.html (hunk 22 above)
Comment (marcoscaceres) [inline index.html]: Do you mean as part of the task?

Comment (marcoscaceres) [inline index.html]: Ok, yes. What I'm not sure of is if events fire before resolving the promise or after.

File: index.html (hunk 23 above)
Comment (marcoscaceres) [inline index.html]: hmmm... ok, this might require a bit of work then. Need to somehow perform this comparison but I don't know if we have everything we need (as inputs) to do that in the algorithm.

File: index.html (hunk 14 above)
Comment (mounirlamouri) [inline index.html]: I think you missed that comment.

**Comment by kenchris** [comment]
@mounirlamouri comment in the files view instead and they won't disappear when force pushed
